include/dali/core/float16.h

```diff
@@ -61,22 +313,32 @@ struct is_fp_or_half {
     std::is_floating_point<T>::value || is_half<T>::value;
 };
 
-// If not clang, just let everyone deal with the alias
-#if defined(__clang__) && defined(__CUDA__)
-template <typename T>
-using to_gpu_t = std::conditional_t<is_half<T>::value, __half, T>;
-
-#define DALI_TO_GPU_T(T) to_gpu_t<T>
+}  // namespace dali
 
+DALI_HOST_DEV DALI_FORCEINLINE
+dali::float16 fma(dali::float16 a, dali::float16 b, dali::float16 c) noexcept {
+#ifdef __CUDA_ARCH__
+#if __CUDA_ARCH__ >= 530
+  return __hfma(a.impl, b.impl, c.impl);
 #else
-
-template <typename T>
-using to_gpu_t = T;
-
-#define DALI_TO_GPU_T(T) T
-
+  return dali::float16(fmaf(static_cast<float>(a.impl),
+                            static_cast<float>(b.impl),
+                            static_cast<float>(c.impl)));
 #endif
+#else
+  return static_cast<float>(a) * static_cast<float>(b) + static_cast<float>(c);
+#endif
+}
 
-}  // namespace dali
+#if !defined(__CUDA_ARCH__) || __CUDA_ARCH__ >= 350  // this is for clang-only build
```

Comment:
Note: clang-only build needs `__device__` function prototypes when compiling host-only code. In NVCC build, this whole function could be removed under the same condition.